File: .net & mono/Weed3/IQuery.cs
using Noear.Weed.Cache;
using System;
using System.Collections.Generic;

namespace Noear.Weed {
    public interface IQuery {
        Object getValue();
        T getValue<T>(T def);

        Variate getVariate();
        Variate getVariate(Action<CacheUsing, Variate> cacheCondition);

        T getItem<T>(T model) where T : class, IBinder;
        T getItem<T>(T model, Action<CacheUsing, T> cacheCondition) where T : class, IBinder;

        List<T> getList<T>(T model) where T : class, IBinder;
        List<T> getList<T>(T model, Action<CacheUsing, List<T>> cacheCondition) where T : class, IBinder;

        DataList getDataList();
        DataList getDataList(Action<CacheUsing, DataList> cacheCondition);
        DataItem getDataItem();
        DataItem getDataItem(Action<CacheUsing, DataList> cacheCondition);

        List<T> getArray<T>(String column);

        IQuery caching(ICacheService service);
        IQuery usingCache(bool isCache);
        IQuery usingCache(int seconds);
        IQuery cacheTag(String tag);
    }
}

File: weed3/src/main/java/org/noear/weed/IQuery.java
package org.noear.weed;

import org.noear.weed.cache.CacheUsing;
import org.noear.weed.cache.ICacheController;
import org.noear.weed.cache.ICacheService;
import org.noear.weed.ext.Act2;

import java.sql.SQLException;
import java.util.List;
import java.util.Map;

/**
 * Created by noear on 14/11/12.
 */
public interface IQuery extends ICacheController<IQuery> {
     long getCount() throws SQLException;
     Object getValue() throws SQLException;
     <T> T getValue(T def) throws SQLException;

     Variate getVariate() throws SQLException;
     Variate getVariate(Act2<CacheUsing,Variate> cacheCondition) throws SQLException;

     <T extends IBinder> T getItem(T model) throws SQLException;
     <T extends IBinder> T getItem(T model, Act2<CacheUsing, T> cacheCondition) throws SQLException;

     <T extends IBinder> List<T> getList(T model) throws SQLException;
     <T extends IBinder> List<T> getList(T model, Act2<CacheUsing, List<T>> cacheCondition) throws SQLException;

     <T> T getItem(Class<T> cls) throws SQLException;
     <T> T getItem(Class<T> cls,Act2<CacheUsing, T> cacheCondition) throws SQLException;

     <T> List<T> getList(Class<T> cls) throws SQLException;
     <T> List<T> getList(Class<T> cls,Act2<CacheUsing, List<T>> cacheCondition) throws SQLException;

     DataList getDataList() throws SQLException;
     DataList getDataList(Act2<CacheUsing, DataList> cacheCondition) throws SQLException;
     DataItem getDataItem() throws SQLException;
     DataItem getDataItem(Act2<CacheUsing, DataItem> cacheCondition) throws SQLException;

     List<Map<String,Object>> getMapList() throws SQLException;
     Map<String,Object> getMap() throws SQLException;

     <T> List<T> getArray(String column) throws SQLException;
     <T> List<T> getArray(int columnIndex) throws SQLException;
     default  <T> List<T> getArray() throws SQLException{
          return getArray(0);
     }
}
